### PR TITLE
Order sanity checks

### LINKED ISSUE #42: Updates streaming fails without recovery
```
websocket_server_3     | 2018-02-18T01:06:25.110355343Z asyncio      ERROR    Task exception was never retrieved
websocket_server_3     | 2018-02-18T01:06:25.110380617Z future: <Task finished coro=<stream_updates() done, defined at /usr/src/app/app/services/websocket_server.py:364> exception=InvalidOperation([<class 'decimal.DivisionUndefined'>],)>
websocket_server_3     | 2018-02-18T01:06:25.110384568Z Traceback (most recent call last):
websocket_server_3     | 2018-02-18T01:06:25.110386780Z   File "/usr/src/app/app/services/websocket_server.py", line 374, in stream_updates
websocket_server_3     | 2018-02-18T01:06:25.110389149Z     "buys": [format_order(order) for order in orders_buys],
websocket_server_3     | 2018-02-18T01:06:25.110391362Z   File "/usr/src/app/app/services/websocket_server.py", line 374, in <listcomp>
websocket_server_3     | 2018-02-18T01:06:25.110393782Z     "buys": [format_order(order) for order in orders_buys],
websocket_server_3     | 2018-02-18T01:06:25.110395923Z   File "/usr/src/app/app/services/websocket_server.py", line 231, in format_order
websocket_server_3     | 2018-02-18T01:06:25.110398138Z     price = base_contract.denormalize_value(record["amount_give"]) / coin_contract.denormalize_value(record["amount_get"])
websocket_server_3     | 2018-02-18T01:06:25.110400560Z decimal.InvalidOperation: [<class 'decimal.DivisionUndefined'>]
```

### PR DESCRIPTION
**Pull request by mmhh1910**
* Changed cerberus order validation for amountGet and amountGive to ">=1" (was >=0).
* Added cerberus  validation so that "tokenGet != tokenGive". (Trap orders for poorly written bots ending up buying 10 X for 100 X)
* Enabled cerberus validation for orders coming from ED (with slightly different rules)
* Added expiration and signature validation for orders coming from ED

This should fix #42 and #45.

### CODE REVIEW

File: app/services/etherdelta_observer.py [hunk 1]
```diff
@@ -41,14 +43,36 @@
...
+
+
 async def process_orders(orders):
     not_deleted_filter = lambda order: "deleted" not in order or not order["deleted"]
     logger.info("Processing %i orders", len(orders))
     orders = list(filter(not_deleted_filter, orders))
     logger.debug("Filtered orders: %i", len(orders))
-
+    v = OrderMessageValidatorEtherdelta()
+    warnings.filterwarnings("ignore", category=DeprecationWarning)
```
Comment (freeatnet) [inline app/services/etherdelta_observer.py]: Does this actually help?

Comment (mmhh1910) [reply app/services/etherdelta_observer.py]: Oh rats... Forgot to delete that line. It does not help, so I deleted the "import warnings", but forgot this line.

Comment (freeatnet) [reply app/services/etherdelta_observer.py]: Ok, this definitely breaks things now.

File: app/services/etherdelta_observer.py [hunk 2]
```diff
@@ -41,14 +43,36 @@
...
+
+            if order_validated["expires"] <= current_block:
+                error_msg = "Cannot post order because it has already expired"
+                details_dict = { "blockNumber": current_block, "expires": order["expires"], "date": datetime.utcnow() }
+                logger.warning("ED Order rejected: %s: %s", error_msg, details_dict)
+                continue
+
+            if not order_signature_valid(order_validated):
+                logger.warning("ED Order rejected: Invalid signature: order = %s", order)
+                continue
```
Comment (freeatnet) [inline app/services/etherdelta_observer.py]: The validation process feels like a candidate for a `filter` predicate function to me. What do you think?

**Review by freeatnet — APPROVED** [review]
Thank you for your contribution, @betterdelta! This is great.

I have a small question for you, but this otherwise looks good to go.

File: app/services/etherdelta_observer.py (hunk 2 above)
Comment (mmhh1910) [reply app/services/etherdelta_observer.py]: I think it's a matter of personal coding style. 
For me the separate validations inside the for loop are readable and easy to understand. I'm fine reading the same with a lambda function, but I tend to not write it like that. 
Go ahead and change it if you want.

Comment (freeatnet) [reply app/services/etherdelta_observer.py]: Didn't mean to imply using a lambda; rather, just extract it into a function that returns False where we currently have `continue` and True at the end?

Comment (mmhh1910) [reply app/services/etherdelta_observer.py]: I don't understand. Just go ahead and make the change.

File: app/src/order_signature.py [hunk 3]
```diff
@@ -15,12 +15,12 @@ def order_signature_valid(message):
...
 
-    signature_base = Web3.sha3(Web3.toHex(eth_signature_prefix + hash_bytes))
+    signature_base = Web3.sha3(hexstr=Web3.toHex(eth_signature_prefix + hash_bytes))
 
     recovered_address = ecrecover(
         Web3.toBytes(hexstr=signature_base),
         message["v"],
-        Web3.toInt(message["r"]),
-        Web3.toInt(message["s"]))
+        Web3.toInt(hexstr=Web3.toHex(message["r"])), # Convert from bytes to hex because sending bytes to toInt is deprecated behaviour, apparently
```
Comment (freeatnet) [inline app/src/order_signature.py:23]: Thanks for making me look into this and fix deprecation warnings, @betterdelta!

File: app/services/etherdelta_observer.py (hunk 2 above)
Comment (freeatnet) [reply app/services/etherdelta_observer.py]: Check out 55ce3e6.

**Comment by freeatnet** [comment]
Seriously awesome stuff, @betterdelta! Thank you for your contribution and I look forward to more PRs from you :)

**Pull request merged**